client/engine/store/persiststore.go

```diff
@@ -0,0 +1,474 @@
+
+// DestroyChannel deletes the channel with id id.
+func (ps *PersistStore) DestroyConsensusChannel(id types.Destination) {
+	err := ps.consensusChannels.Update(func(tx *buntdb.Tx) error {
+		_, err := tx.Delete(id.String())
+		return err
+	})
+	ps.checkError(err)
+}
+
+// GetChannelById retrieves the channel with the supplied id, if it exists.
+func (ps *PersistStore) GetChannelById(id types.Destination) (c *channel.Channel, ok bool) {
+	ch, err := ps.getChannelById(id)
+
+	if err != nil {
+		return &channel.Channel{}, false
+	}
+
+	return &ch, true
+}
+
+// getChannelById returns the stored channel
+func (ps *PersistStore) getChannelById(id types.Destination) (channel.Channel, error) {
+	var chJSON string
+	err := ps.channels.View(func(tx *buntdb.Tx) error {
+		var err error
+		chJSON, err = tx.Get(id.String())
+		return err
+	})
+
+	if errors.Is(err, buntdb.ErrNotFound) {
+
+		return channel.Channel{}, ErrNoSuchChannel
+	}
+	var ch channel.Channel
+	err = ch.UnmarshalJSON([]byte(chJSON))
+
+	if err != nil {
+		return channel.Channel{}, fmt.Errorf("error unmarshaling channel %s", ch.Id)
+	}
+
+	return ch, nil
+}
+
+// GetChannelsByParticipant returns any channels that include the given participant
+func (ps *PersistStore) GetChannelsByParticipant(participant types.Address) []*channel.Channel {
+	toReturn := []*channel.Channel{}
+	err := ps.channels.View(func(tx *buntdb.Tx) error {
+		err := tx.Ascend("", func(key, chJSON string) bool {
+
+			var ch channel.Channel
+			err := json.Unmarshal([]byte(chJSON), &ch)
+
+			if err != nil {
+				return true // channel not found, continue looking
+			}
+
+			participants := ch.FixedPart.Participants
+			for _, p := range participants {
+				if p == participant {
+					toReturn = append(toReturn, &ch)
+				}
+
+			}
+
+			return true // channel not found: continue looking
+		})
+		return err
+	})
+	ps.checkError(err)
+	return toReturn
+}
+
+// GetConsensusChannelById returns a ConsensusChannel with the given channel id
+func (ps *PersistStore) GetConsensusChannelById(id types.Destination) (channel *consensus_channel.ConsensusChannel, err error) {
+
+	var ch *consensus_channel.ConsensusChannel
+	err = ps.consensusChannels.View(func(tx *buntdb.Tx) error {
+
+		chJSON, err := tx.Get(id.String())
+
+		if errors.Is(err, buntdb.ErrNotFound) {
+			return ErrNoSuchChannel
+		}
+
+		ch = &consensus_channel.ConsensusChannel{}
+		err = ch.UnmarshalJSON([]byte(chJSON))
+
+		if err != nil {
+			return fmt.Errorf("error unmarshaling channel %s", ch.Id)
+		}
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+	return ch, nil
+}
+
+// GetConsensusChannel returns a ConsensusChannel between the calling client and
+// the supplied counterparty, if such channel exists
+func (ps *PersistStore) GetConsensusChannel(counterparty types.Address) (channel *consensus_channel.ConsensusChannel, ok bool) {
+
+	err := ps.consensusChannels.View(func(tx *buntdb.Tx) error {
+		return tx.Ascend("", func(key, chJSON string) bool {
```

Comment:
These nested passed functions are a bit of a stretch for my conceptual capacities, but I don't have any concrete suggestions for simplifying.